lib/cdo/rack/attack.rb

```diff
@@ -0,0 +1,59 @@
+require 'rack/attack'
+require 'active_support/core_ext/numeric/time'
+
+class Rack::Attack
+  # maximum requests per minute over a 4-minute stretch.
+  MAX_TABLE_READS_PER_MIN = 2 * 60
+  MAX_TABLE_WRITES_PER_MIN = 4 * 60
+  MAX_PROPERTY_READS_PER_MIN = 4 * 60
+  MAX_PROPERTY_WRITES_PER_MIN = 4 * 60
+
+  redis_url = CDO.geocoder_redis_url || 'redis://localhost:6379'
+  cache.store = StoreProxy::RedisStoreProxy.new(Redis.new(url: redis_url))
+
+  def self.limits(max_per_min)
+    # allow 4x the max rate over 15s, 2x the max rate over 60s, and 1x the max rate over 4min.
+    # this gives the experience of exponential backoff when limits are exceeded.
+    [[max_per_min, 15],
+     [2 * max_per_min, 60],
+     [4 * max_per_min, 240]]
+  end
+
+  class Request < ::Rack::Request
```

Comment:
It seems to work without the `::`. In fact the request seems to know its superclass without specifying it here at all. However, `::Rack` tells us at a glance that we are using the top-level Rack module without having to read through the RackAttack module to verify that there is no other Rack::Request we might be picking up. So I guess I'll leave it this way for "readability".

FWIW it seems like some other API documentation shows `::` before their API calls (e.g. `::NewRelic`). I am guessing it is some kind of insurance policy against users who wrap their call to the API in another module with the same name.